### PR TITLE
feat(ext/kv): key-value store

### PR DESCRIPTION
**Pull request by losfair**
This commit adds unstable "Deno.openKv()" API that allows to open
a key-value database at a specified path.

### CODE REVIEW

File: ext/kv/lib.rs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+  path: Option<String>,
+) -> Result<ResourceId, AnyError>
+where
+  DBH: DatabaseHandler + 'static,
+{
+  let handler = {
+    let state = state.borrow();
+    state.borrow::<Rc<DBH>>().clone()
+  };
+  let db = handler.open(path).await?;
```
Comment (dsherret) [inline ext/kv/lib.rs:86]: I think we need to check --read and --write permissions for this file here? Also, we need unstable checks. You can find some examples in the code by searching for "check_unstable"

File: ext/kv/generate.ts
Comment (crowlKats) [inline ext/kv/generate.ts:1]: why is this necessary?

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 2]
```diff
@@ -1465,6 +1473,94 @@ declare namespace Deno {
    * @category HTTP Server
    */
   export function upgradeHttpRaw(request: Request): [Deno.Conn, Uint8Array];
+
+  /** **Experimental addon API**
+   *
+   * Open a new {@link Deno.Database} connection to persist data.
+   */
```
Comment (crowlKats) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1480]: please use a category for all these APIs

File: ext/kv/generate.ts
Comment (losfair) [reply ext/kv/generate.ts:1]: To generate `01_db.js` from `01_db.ts`.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
Marking as "Request changes" because we're in the middle of refactor to ops in 
https://github.com/denoland/deno/pull/18080 and https://github.com/denoland/deno/pull/18210. 

Let's keep reviewing this PR, but I want to hold of until weekend with landing this one.

File: ext/kv/lib.rs (hunk 1 above)
Comment (bartlomieju) [reply ext/kv/lib.rs:86]: Agreed, we should also have at least one test case that verifies that permissions are checked

File: ext/kv/sqlite.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    let conn = match (path, &self.default_storage_dir) {
+      (Some(path), _) => {
+        let path = PathBuf::from(path);
+        // TODO(@lucacasonato): check permissions
+        rusqlite::Connection::open(path)?
+      }
+      (None, Some(path)) => {
+        std::fs::create_dir_all(path)?;
+        let path = path.join("state.sqlite3");
+        rusqlite::Connection::open(&path)?
```
Comment (bartlomieju) [inline ext/kv/sqlite.rs:100]: Permissions should be checked here too

File: ext/kv/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+description = "Implementation of the Deno database API"
+
+[lib]
+path = "lib.rs"
+
+[dependencies]
+anyhow.workspace = true
+async-trait.workspace = true
+deno_core.workspace = true
+hex = "0.4"
```
Comment (bartlomieju) [inline ext/kv/Cargo.toml:20]: We use `hex` in `ext/node` - could you please lift both of the to root `Cargo.toml`?

File: ext/kv/generate.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+import * as esbuild from "https://deno.land/x/esbuild@v0.14.48/wasm.js";
+
+const source = await Deno.readTextFile("./01_db.ts");
```
Comment (bartlomieju) [inline ext/kv/generate.ts:5]: I suggest to use `import.meta.resolve()` here so this file is not dependent being run from `ext/kv` directory

File: ext/kv/generate.ts
Comment (crowlKats) [reply ext/kv/generate.ts:1]: but why is this necessary? we shouldnt have a build-step for source files.

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 6]
```diff
@@ -163,7 +163,7 @@ declare namespace Deno {
    */
   type ToNativeResultType<T extends NativeResultType = NativeResultType> =
     T extends NativeStructType ? BufferSource
-    : ToNativeResultTypeMap[Exclude<T, NativeStructType>];
+      : ToNativeResultTypeMap[Exclude<T, NativeStructType>];
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply cli/tsc/dts/lib.deno.unstable.d.ts:166]: These are correct changes. `cli/tsc/dts/lib.deno.unstable.d.ts` was previously not included by `./tools/format.ts` (I have fixed this now).

File: ext/kv/generate.ts
Comment (bartlomieju) [reply ext/kv/generate.ts:1]: https://github.com/denoland/deno/blob/e6fe1630f0e6cbc16ccfda5a58cbe8ecdfea2694/runtime/build.rs#L24-L59

Comment (losfair) [reply ext/kv/generate.ts:1]: Yes it indeed works! Updated.

File: ext/kv/lib.rs (hunk 1 above)
Comment (ry) [reply ext/kv/lib.rs]: The default database is treated like localStorage - it can be stored in the DENO_DIR

**Comment by lucacasonato** [comment]
todo:
- [ ] test key serialization from/to js (bigint panics right now)
- [ ] move selector & cursor parsing logic to rust

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 7]
```diff
@@ -1465,6 +1473,505 @@ declare namespace Deno {
...
+   * When no path is provided, the database will be opened in a default path for
+   * the current script. This location is persistent across script runs and is
+   * keyed on the origin storage key (the same key that is used to determine
+   * `localStorage` persistence). More information about the origin storage key
+   * can be found in the Deno Manual.
+   *
+   * @tags allow-read, allow-write
+   * @category KV
+   */
+  export function openDatabase(path?: string): Promise<Deno.Database>;
```
Comment (ry) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1493]: Rename to `Deno.kv()`

**Review by bartlomieju — APPROVED** [review]
LGTM, I read through the source and besides some nitpicks there's nothing standing out.

We can address these nitpicks after the release.

**Comment by jsejcksn** [comment]
@losfair I see a [`Deno.Kv#getMany`](https://deno.land/api@v1.32.0?unstable=&s=Deno.Kv#method_getMany_0) method in the types, but a corresponding method does not exist on an instance at runtime. Was the method a short-lived experiment or am I missing something?

`example.ts`:
```ts
/// <reference lib="deno.unstable" />

const db = await Deno.openKv(":memory:");

console.log(Object.getOwnPropertyNames(Object.getPrototypeOf(db)));

console.log(await db.getMany([]));

await db.close();

```

```
% deno --version
deno 1.32.0 (release, aarch64-apple-darwin)
v8 11.2.214.9
typescript 5.0.2

% deno check example.ts

% echo $?
0

% deno run --unstable example.ts
[
  "constructor",
  "atomic",
  "get",
  "set",
  "delete",
  "list",
  "close"
]
error: Uncaught TypeError: db.getMany is not a function
console.log(await db.getMany([]));
                     ^
    at file:///Users/deno/example.ts:12:22
```
